### PR TITLE
Energy saving [3.6.0-beta]

### PR DESCRIPTION
**Pull request by hannesa2**

### CODE REVIEW

File: serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt [hunk 1]
```diff
@@ -2,154 +2,55 @@ package info.mqtt.android.service.ping
...
     override fun start() {
-        val action = MqttServiceConstants.PING_SENDER + clientComms!!.client.clientId
-        Timber.d("Register AlarmReceiver to MqttService$action")
-        service.registerReceiver(alarmReceiver, IntentFilter(action))
-        pendingIntent = PendingIntent.getBroadcast(service, 0, Intent(action), PendingIntent.FLAG_UPDATE_CURRENT)
-        schedule(clientComms!!.keepAlive)
-        hasStarted = true
+        val pingRepeatWorkRequest = PeriodicWorkRequest
+            .Builder(PingWorker::class.java, clientComms!!.keepAlive / 100, TimeUnit.MILLISECONDS)
+            .addTag(PING_WORKER_TAG)
```
Comment (sirambd) [inline serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt:33]: You add a tag just to be able to cancel the work, but except that it is unique, so you can cancel it with the unique work name

File: serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt [hunk 2]
```diff
@@ -2,154 +2,55 @@ package info.mqtt.android.service.ping
...
-                alarmManager.cancel(pendingIntent)
-            }
-            hasStarted = false
-            try {
-                service.unregisterReceiver(alarmReceiver)
-            } catch (e: IllegalArgumentException) {
-                //Ignore unregister errors.
-            }
-        }
+        workManager.cancelAllWorkByTag(PING_WORKER_TAG)
```
Comment (sirambd) [inline serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt:43]: Just `cancelUniqueWork(PING_JOB)` it works

**Review by sirambd — COMMENTED** [review]
Quick review, i hope you will enjoy 😁

File: serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt (hunk 2 above)
Comment (hannesa2) [reply serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt]: Is this the desired change 2692bc87 ?

Comment (sirambd) [reply serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt]: Yes 👍🏽

File: serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt [hunk 3]
```diff
@@ -1,155 +1,52 @@
...
 
     override fun start() {
-        val action = MqttServiceConstants.PING_SENDER + clientComms!!.client.clientId
-        Timber.d("Register AlarmReceiver to MqttService$action")
-        service.registerReceiver(alarmReceiver, IntentFilter(action))
-        pendingIntent = PendingIntent.getBroadcast(service, 0, Intent(action), PendingIntent.FLAG_UPDATE_CURRENT)
-        schedule(clientComms!!.keepAlive)
-        hasStarted = true
+        val pingRepeatWorkRequest = PeriodicWorkRequest
+            .Builder(PingWorker::class.java, clientComms!!.keepAlive / 100, TimeUnit.MILLISECONDS)
```
Comment (sirambd) [inline serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt:31]: I will just have a quick question, why divide `keepAlive` by 100?

Comment (hannesa2) [reply serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt:31]: Indeed, that makes no sense and is a bug. 👍
image

**Comment by hannesa2** [comment]
There is a [beta version](https://github.com/hannesa2/paho.mqtt.android/releases/tag/3.4.0-beta1) containing this feature

`implementation 'com.github.hannesa2:paho.mqtt.android:3.4.0-beta1'`

I'm happy when someone will test it and give me feedback. Otherwise I hesitate to merge it

File: extendedSample/src/main/java/info/mqtt/android/extsample/internal/Connection.kt [hunk 4]
```diff
@@ -196,7 +196,7 @@ class Connection private constructor(
     }
 
     companion object {
-        private const val FOREGROUND = true
+        private const val FOREGROUND = false
```
Comment (rohittayal96) [inline extendedSample/src/main/java/info/mqtt/android/extsample/internal/Connection.kt:199]: Why was this variable changed?

Comment (rohittayal96) [inline extendedSample/src/main/java/info/mqtt/android/extsample/internal/Connection.kt:199]: Why was this variable changed?

Comment (hannesa2) [reply extendedSample/src/main/java/info/mqtt/android/extsample/internal/Connection.kt:199]: This was on of the main goals to remove it. There is no need of a foreground service anymore

**Comment by hantrungkien** [comment]
```
E/WM-WorkerFactory: Could not instantiate info.mqtt.android.service.ping.AlarmPingSender$PingWorker
    java.lang.NoSuchMethodException: info.mqtt.android.service.ping.AlarmPingSender$PingWorker.<init> [class android.content.Context, class androidx.work.WorkerParameters]
```

@hannesa2 do you meet this same problem?

**Comment by hantrungkien** [comment]
@hannesa2 no bro, I have found it on my product which is using Koin for DI and also KoinWorkerFactory. I think may be that is reason.

**Comment by hannesa2** [comment]
@hantrungkien When you use this beta version, please let me know how it works for you. Thank you !

File: extendedSample/src/main/java/info/mqtt/android/extsample/internal/Connection.kt (hunk 4 above)
Comment (hannesa2) [reply extendedSample/src/main/java/info/mqtt/android/extsample/internal/Connection.kt:206]: There is no need anymore. This is the goal of this chance

**Comment by hannesa2** [comment]
close https://github.com/hannesa2/paho.mqtt.android/issues/230

**Comment by hannesa2** [comment]
related to https://github.com/hannesa2/paho.mqtt.android/issues/37

**Comment by hannesa2** [comment]
There is a new beta version with these changes https://github.com/hannesa2/paho.mqtt.android/releases/tag/3.6.0-beta4

**Comment by hannesa2** [comment]
I continue with #430
Please follow the progress there

**Pull request closed without merging**